src/test/test-message-digest.c

```diff
@@ -0,0 +1,405 @@
+    struct chunked_ctx *ctx = data;
+    uint8_t i;
+
+    if (ctx->offset >= ctx->t->len)
+        return false;
+
+    /* feed 3 blobs from within the same main loop iteration, then
+     * wait to be completed and send more.
+     */
+    for (i = 0; i < 3; i++) {
+        struct sol_blob *blob;
+        bool is_final = false;
+        size_t len = CHUNK_SIZE;
+        const void *mem;
+        int r;
+
+        if (ctx->offset + CHUNK_SIZE >= ctx->t->len) {
+            is_final = true;
+            len = ctx->t->len - ctx->offset;
+        }
+
+        mem = (char *)ctx->t->mem + ctx->offset;
+        blob = sol_blob_new(SOL_BLOB_TYPE_NOFREE, NULL, mem, len);
+        ASSERT(blob != NULL);
+
+        ctx->offset += len;
+
+        r = sol_message_digest_feed(ctx->md, blob, is_final);
+        ASSERT_INT_EQ(r, 0);
+
+        sol_blob_unref(blob);
+
+        if (is_final)
+            break;
+    }
+
+    /* keep calling this function (will call from different main loop
+     * iteration (10ms timer) and possibly before on_feed_done.
+     */
+    return ctx->offset < ctx->t->len;
+}
+
+static void
+on_feed_done_chunked(void *data, struct sol_message_digest *handle, struct sol_blob *input)
+{
+    /* feed more after we're done with our previous data */
+    on_timeout_do_chunked_internal(data);
+}
+
+static void
+on_digest_ready_chunked(void *data, struct sol_message_digest *handle, struct sol_blob *digest)
+{
+    struct chunked_ctx *ctx = data;
+
+    on_digest_ready_simple((void *)ctx->t, handle, digest);
+    free(ctx);
+}
+
+static bool
+on_timeout_do_chunked(void *data)
+{
+    const struct md_test *itr = data;
+
+    for (; itr->mem != NULL; itr++) {
+        struct chunked_ctx *ctx;
+        struct sol_message_digest_config cfg = {
+            .api_version = SOL_MESSAGE_DIGEST_CONFIG_API_VERSION,
+            .algorithm = itr->algorithm,
+            .key = itr->key,
+            .on_digest_ready = on_digest_ready_chunked,
+            .on_feed_done = on_feed_done_chunked,
+        };
+
+        ctx = malloc(sizeof(*ctx));
+        ASSERT(ctx != NULL);
+        ctx->t = itr;
+        ctx->offset = 0;
+        cfg.data = ctx;
+
+        ctx->md = sol_message_digest_new(&cfg);
+        ASSERT(ctx->md != NULL);
+
+        sol_timeout_add(10, on_timeout_do_chunked_internal, ctx);
+
+        pending++;
+    }
+
+    return false;
+}
+
+DEFINE_TEST(test_md5_single);
+
+static void
+test_md5_single(void)
+{
+    /* note: mem includes string trailing null byte */
+    static const struct md_test md5_single_test[] = {
+        MD_TEST("md5", SOL_STR_SLICE_EMPTY, "test", "e2a3e68d23ce348b8f68b3079de3d4c9"),
+        MD_TEST("md5", SOL_STR_SLICE_EMPTY, "long line of text bla bla bla more text here yada yada", "18511ce4f220de4744390ca3ae72873f"),
+        MD_TEST("md5", SOL_STR_SLICE_EMPTY, big_blob_of_zeros, "ab893875d697a3145af5eed5309bee26"),
+        MD_TEST("md5", SOL_STR_SLICE_EMPTY, big_blob_of_chars, "9a36eacb09f8e98e103e9ee897f8e31c"),
```

Comment:
I"d also test for the empty string (which should evaluate to `d41d8cd98f00b204e9800998ecf8427e`).